### PR TITLE
Detect .hs-boot files

### LINKED ISSUE #22: Support hs-boot files
`gazelle_cabal` does not pick up hs-boot files currently. These should be listed in the `srcs` attribute of the generated rules when present.

### PR DESCRIPTION
**Pull request by googleson78**
Fixes #22

### CODE REVIEW

File: example/package-a/src/PackageA/Other/LoopBoot.hs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module PackageA.Other.LoopBoot where
+
+import LoopNonBoot (LoopNonBoot(..))
```
Comment (facundominguez) [inline example/package-a/src/PackageA/Other/LoopBoot.hs:3]: Maybe this should be `import PackageA.Other.LoopNonBoot (LoopNonBoot(..))` ?

Comment (googleson78) [reply example/package-a/src/PackageA/Other/LoopBoot.hs]: Yup, thanks

**Comment by googleson78** [comment]
I've converted this to a draft until I can confirm it works locally. Sorry for the wasted time.

File: cabalscan/src/CabalScan/RuleGenerator.hs [hunk 2]
```diff
@@ -281,6 +284,20 @@ depPackageNames = concatMap depNames . Cabal.targetBuildDepends
         in
           map identifierOf $ Set.toList $ Cabal.depLibraries dep
 
+data FoundModulePath = FoundModulePath
```
Comment (facundominguez) [inline cabalscan/src/CabalScan/RuleGenerator.hs:287]: This type is requiring a module to be put together with its boot module. Is this necessary for any further processing? As far as I can see this link is discarded in `foundModulePathToPathList` almost immediately after going through the trouble of putting it together.

File: defs.bzl [hunk 3]
```diff
@@ -22,6 +24,7 @@ _gazelle_cabal_dependencies = repository_rule(
     local = True,
     attrs = {
         "aeson": attr.label(default = "@stackage//:aeson"),
+        "exceptions": attr.label(default = "@stackage//:exceptions"),
```
Comment (facundominguez) [inline defs.bzl:27]: Adding dependencies could require the user to be aware of them. It would be better to avoid it if it isn't essential.

**Review by facundominguez — COMMENTED** [review]
Thanks @googleson78!

I added some comments. Perhaps we can simplify/reduce a bit this PR ahead of merging it.

File: defs.bzl (hunk 3 above)
Comment (googleson78) [reply defs.bzl:27]: I agree, I'll drop this.

Comment (googleson78) [reply defs.bzl]: Done

File: cabalscan/src/CabalScan/RuleGenerator.hs (hunk 2 above)
Comment (facundominguez) [reply cabalscan/src/CabalScan/RuleGenerator.hs:285]: Seems that boot files are ignored if no corresponding module is present, but this is more behavior than has been discussed so far. Is this why you want to return either a 2-tuple or a `FoundModulePath`? I still don't get it :)

Comment (googleson78) [reply cabalscan/src/CabalScan/RuleGenerator.hs:285]: Yep, if I only return the first module I match with an extension, either only the `.hs` file or only the `.hs-boot` file will match, but not both. So we need to return both of them. Either a tuple, or a list, or a record. I prefer the record approach here, more clear than a tuple.

Comment (facundominguez) [reply cabalscan/src/CabalScan/RuleGenerator.hs:285]: Thanks for the elaboration. That starts to make sense.

Now, note that boot files can end with extension `.lhs-boot` as well. And I'm realizing that we would need to define the behavior when both `.lhs-boot` and `.hs-boot` files are present. And similarly in older code, when ".hs" and ".lhs" files are both present.

In light of this, how about we just collect all files matching any of the extensions ".hs", ".lhs", ".hs-boot", ".lhs-boot", and ".hsc"?

We could leave it to `rules_haskell` to fail if any of file combinations is incompatible.

Comment (googleson78) [reply cabalscan/src/CabalScan/RuleGenerator.hs:285]: Good point, thanks.

Comment (googleson78) [reply cabalscan/src/CabalScan/RuleGenerator.hs]: Done

**Review by facundominguez — APPROVED** [review]
Thanks @googleson78!

**Pull request merged**